File: seema-api/routers/email_admin.py
"""Email admin routes — settings, templates, queue management."""
import uuid
import json
from datetime import datetime, timezone
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, File, UploadFile
from pydantic import BaseModel
from sqlalchemy import select, update, func
from sqlalchemy.ext.asyncio import AsyncSession

from middleware.tenant_rls import tenant_db_from_jwt, bypass_db
from middleware.auth import get_current_user, CurrentUser
from services.audit_logger import log_audit
from models.email import EmailTemplate, EmailQueueItem
from models.firm import Firm

router = APIRouter()

class EmailSettings(BaseModel):
    sender_email: str
    sender_name: str
    smtp_server: Optional[str] = None
    smtp_port: Optional[int] = None
    enable_auto_chase: bool = False
    chase_frequency_days: int = 7
    max_retries: int = 3

class TestEmailRequest(BaseModel):
    recipient: str
    subject: Optional[str] = "Test Email from Seema"

@router.get("/admin/email-settings")
async def get_email_settings(current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    result = await db.execute(select(Firm).where(Firm.id == current_user.firm_id))
    firm = result.scalar_one_or_none()
    if not firm:
        raise HTTPException(status_code=404, detail="Firm not found")
    prefs = {}
    if firm.firm_preferences:
        try:
            prefs = json.loads(firm.firm_preferences)
        except:
            prefs = {}
    email_settings = prefs.get("email_settings", {"sender_email": "", "sender_name": "", "smtp_server": None, "smtp_port": None, "enable_auto_chase": False, "chase_frequency_days": 7, "max_retries": 3})
    return email_settings

@router.post("/admin/email-settings")
async def save_email_settings(settings: EmailSettings, current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    result = await db.execute(select(Firm).where(Firm.id == current_user.firm_id))
    firm = result.scalar_one_or_none()
    if not firm:
        raise HTTPException(status_code=404, detail="Firm not found")
    prefs = {}
    if firm.firm_preferences:
        try:
            prefs = json.loads(firm.firm_preferences)
        except:
            prefs = {}
    prefs["email_settings"] = settings.dict()
    await db.execute(update(Firm).where(Firm.id == current_user.firm_id).values(firm_preferences=json.dumps(prefs), updated_at=datetime.now(timezone.utc)))
    await db.flush()
    await log_audit(db=db, firm_id=current_user.firm_id, action="email_settings_updated", entity_type="firm", entity_id=current_user.firm_id, user_id=current_user.user_id, details=json.dumps(settings.dict()))
    return {"message": "Email settings saved", "settings": settings.dict()}

@router.get("/admin/email-templates")
async def list_email_templates(current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    result = await db.execute(select(EmailTemplate).where(EmailTemplate.firm_id == current_user.firm_id).order_by(EmailTemplate.created_at.desc()))
    templates = result.scalars().all()
    return [{"id": t.id, "name": t.name, "subject": t.subject, "body": t.body, "category": t.category, "is_active": t.is_active, "created_at": t.created_at.isoformat() if t.created_at else None} for t in templates]

@router.get("/admin/email-queue")
async def list_email_queue(status: Optional[str] = None, limit: int = 100, current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    query = select(EmailQueueItem).where(EmailQueueItem.firm_id == current_user.firm_id)
    if status:
        query = query.where(EmailQueueItem.status == status)
    query = query.order_by(EmailQueueItem.created_at.desc()).limit(limit)
    result = await db.execute(query)
    items = result.scalars().all()
    return [{"id": item.id, "firm_id": item.firm_id, "recipient": item.recipient, "subject": item.subject, "status": item.status, "sent_at": item.sent_at.isoformat() if item.sent_at else None, "created_at": item.created_at.isoformat() if item.created_at else None} for item in items]

@router.get("/admin/email-queue/stats")
async def email_queue_stats(current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    total_result = await db.execute(select(func.count(EmailQueueItem.id)).where(EmailQueueItem.firm_id == current_user.firm_id))
    total = total_result.scalar() or 0
    pending_result = await db.execute(select(func.count(EmailQueueItem.id)).where((EmailQueueItem.firm_id == current_user.firm_id) & (EmailQueueItem.status == "pending")))
    pending = pending_result.scalar() or 0
    sent_result = await db.execute(select(func.count(EmailQueueItem.id)).where((EmailQueueItem.firm_id == current_user.firm_id) & (EmailQueueItem.status == "sent")))
    sent = sent_result.scalar() or 0
    failed_result = await db.execute(select(func.count(EmailQueueItem.id)).where((EmailQueueItem.firm_id == current_user.firm_id) & (EmailQueueItem.status == "failed")))
    failed = failed_result.scalar() or 0
    bounced_result = await db.execute(select(func.count(EmailQueueItem.id)).where((EmailQueueItem.firm_id == current_user.firm_id) & (EmailQueueItem.status == "bounced")))
    bounced = bounced_result.scalar() or 0
    return {"total": total, "pending": pending, "sent": sent, "failed": failed, "bounced": bounced}

@router.post("/admin/email-queue/send-all")
async def send_all_pending(current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    result = await db.execute(select(EmailQueueItem).where((EmailQueueItem.firm_id == current_user.firm_id) & (EmailQueueItem.status == "pending")))
    pending_items = result.scalars().all()
    sent_count = 0
    for item in pending_items:
        await db.execute(update(EmailQueueItem).where(EmailQueueItem.id == item.id).values(status="sent", sent_at=datetime.now(timezone.utc)))
        sent_count += 1
    await db.flush()
    await log_audit(db=db, firm_id=current_user.firm_id, action="email_batch_send", entity_type="email_queue", user_id=current_user.user_id, details=json.dumps({"count": sent_count}))
    return {"message": f"{sent_count} emails marked as sent"}

@router.post("/admin/email-queue/{item_id}/send")
async def send_single_email(item_id: str, current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    result = await db.execute(select(EmailQueueItem).where((EmailQueueItem.id == item_id) & (EmailQueueItem.firm_id == current_user.firm_id)))
    item = result.scalar_one_or_none()
    if not item:
        raise HTTPException(status_code=404, detail="Email queue item not found")
    await db.execute(update(EmailQueueItem).where(EmailQueueItem.id == item_id).values(status="sent", sent_at=datetime.now(timezone.utc)))
    await db.flush()
    await log_audit(db=db, firm_id=current_user.firm_id, action="email_send", entity_type="email_queue", entity_id=item_id, user_id=current_user.user_id, details=json.dumps({"recipient": item.recipient}))
    return {"message": "Email marked as sent", "item_id": item_id}

@router.post("/admin/email/test")
async def send_test_email(request: TestEmailRequest, current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    test_email = EmailQueueItem(id=str(uuid.uuid4()), firm_id=current_user.firm_id, recipient=request.recipient, subject=request.subject or "Test Email from Seema", status="pending")
    db.add(test_email)
    await db.flush()
    await log_audit(db=db, firm_id=current_user.firm_id, action="test_email_created", entity_type="email_queue", entity_id=test_email.id, user_id=current_user.user_id, details=json.dumps({"recipient": request.recipient}))
    return {"message": "Test email queued", "item_id": test_email.id, "recipient": request.recipient}

@router.post("/admin/email/auto-chase")
async def trigger_auto_chase(current_user: CurrentUser = Depends(get_current_user), db: AsyncSession = Depends(tenant_db_from_jwt)):
    current_user.require_role("admin")
    sent_count = 0
    await log_audit(db=db, firm_id=current_user.firm_id, action="auto_chase_triggered", entity_type="system", user_id=current_user.user_id, details=json.dumps({"sent_count": sent_count}))
    return {"message": "Auto-chase triggered", "emails_sent": sent_count}

File: seema-node/src/routes/emailAdmin.ts
import { Router, Request, Response } from 'express';
import prisma from '../lib/prisma.js';
import { authenticate, requireRole } from '../middleware/auth.js';
import { getTenantFilter } from '../middleware/tenant.js';
import { logAudit } from '../middleware/auditLogger.js';
import { z } from 'zod';
import sgMail from '@sendgrid/mail';

sgMail.setApiKey(process.env.SENDGRID_API_KEY || '');

const router = Router();

// Serialize Prisma's camelCase to the snake_case shape the frontend expects.
function serializeEmailTemplate(t: any) {
  return {
    id: t.id,
    name: t.name,
    subject: t.subject,
    body: t.body,
    category: t.category,
    created_at: t.createdAt,
    updated_at: t.updatedAt,
  };
}

function serializeEmailQueueItem(i: any) {
  return {
    id: i.id,
    recipient: i.recipient,
    subject: i.subject,
    status: i.status,
    sent_at: i.sentAt,
    error: i.error,
    created_at: i.createdAt,
    updated_at: i.updatedAt,
  };
}

// The email settings POST stores arbitrary JSON in firm preferences. We don't
// know the exact frontend payload shape, so the route still passes through
// req.body verbatim — that's fine because nothing here is validated by Zod
// and the GET returns whatever shape was saved. Leaving as-is.

// GET /admin/email-settings — Return email settings from firm preferences
router.get('/admin/email-settings', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const firm = await prisma.firm.findUnique({ where: { id: firmId } });

    if (!firm) {
      res.status(404).json({ error: true, message: 'Firm not found' });
      return;
    }

    let emailSettings = {
      fromEmail: process.env.SENDGRID_FROM_EMAIL || 'noreply@seemaai.co.uk',
      fromName: firm.name,
      replyTo: firm.email || '',
      autoChaseEnabled: false,
      autoChaseFrequencyDays: 7,
    };

    if (firm.firmPreferences) {
      try {
        const prefs = JSON.parse(firm.firmPreferences);
        if (prefs.emailSettings) {
          emailSettings = { ...emailSettings, ...prefs.emailSettings };
        }
      } catch { /* ignore parse errors */ }
    }

    res.json(emailSettings);
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch email settings' });
  }
});

// POST /admin/email-settings — Save email settings to firm preferences
router.post('/admin/email-settings', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const firm = await prisma.firm.findUnique({ where: { id: firmId } });

    if (!firm) {
      res.status(404).json({ error: true, message: 'Firm not found' });
      return;
    }

    let prefs: Record<string, unknown> = {};
    if (firm.firmPreferences) {
      try {
        prefs = JSON.parse(firm.firmPreferences);
      } catch { /* start fresh */ }
    }

    prefs.emailSettings = req.body;

    await prisma.firm.update({
      where: { id: firmId },
      data: { firmPreferences: JSON.stringify(prefs) },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'email_settings_updated',
      entityType: 'firm',
      entityId: firmId,
    });

    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to save email settings' });
  }
});

// GET /admin/email-templates — List email templates
router.get('/admin/email-templates', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const templates = await prisma.emailTemplate.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
    });
    res.json(templates.map(serializeEmailTemplate));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch email templates' });
  }
});

// GET /admin/email-queue — List email queue items
router.get('/admin/email-queue', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const items = await prisma.emailQueueItem.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
    });
    res.json(items.map(serializeEmailQueueItem));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch email queue' });
  }
});

// GET /admin/email-queue/stats — Count by status
router.get('/admin/email-queue/stats', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const stats = await prisma.emailQueueItem.groupBy({
      by: ['status'],
      where: { firmId },
      _count: { status: true },
    });

    const formatted: Record<string, number> = {};
    for (const item of stats) {
      // status is nullable in the DB — bucket nulls under 'unknown'.
      const key = item.status ?? 'unknown';
      formatted[key] = item._count.status;
    }

    res.json(formatted);
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch email queue stats' });
  }
});

// POST /admin/email-queue/send-all — Process pending emails
router.post('/admin/email-queue/send-all', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const pendingItems = await prisma.emailQueueItem.findMany({
      where: { firmId, status: 'pending' },
    });

    let sent = 0;
    let failed = 0;

    for (const item of pendingItems) {
      try {
        await sgMail.send({
          to: item.recipient,
          from: process.env.SENDGRID_FROM_EMAIL || 'noreply@seemaai.co.uk',
          subject: item.subject,
          text: item.subject, // body from template would be resolved here
        });

        await prisma.emailQueueItem.update({
          where: { id: item.id },
          data: { status: 'sent', sentAt: new Date() },
        });
        sent++;
      } catch (sendErr) {
        await prisma.emailQueueItem.update({
          where: { id: item.id },
          data: {
            status: 'failed',
            error: sendErr instanceof Error ? sendErr.message : 'Unknown error',
          },
        });
        failed++;
      }
    }

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'email_queue_processed',
      entityType: 'email_queue',
      metadata: { total: pendingItems.length, sent, failed },
    });

    res.json({ total: pendingItems.length, sent, failed });
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to process email queue' });
  }
});

// POST /admin/email-queue/:itemId/send — Send single queued email
router.post('/admin/email-queue/:itemId/send', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const { itemId } = req.params as Record<string, string>;

    const item = await prisma.emailQueueItem.findFirst({
      where: { id: itemId, firmId },
    });
    if (!item) {
      res.status(404).json({ error: true, message: 'Email queue item not found' });
      return;
    }

    try {
      await sgMail.send({
        to: item.recipient,
        from: process.env.SENDGRID_FROM_EMAIL || 'noreply@seemaai.co.uk',
        subject: item.subject,
        text: item.subject,
      });

      const updated = await prisma.emailQueueItem.update({
        where: { id: itemId },
        data: { status: 'sent', sentAt: new Date() },
      });

      res.json(serializeEmailQueueItem(updated));
    } catch (sendErr) {
      await prisma.emailQueueItem.update({
        where: { id: itemId },
        data: {
          status: 'failed',
          error: sendErr instanceof Error ? sendErr.message : 'Unknown error',
        },
      });

      res.status(500).json({ error: true, message: 'Failed to send email' });
    }
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to send queued email' });
  }
});

// POST /admin/email/test — Send test email via SendGrid
router.post('/admin/email/test', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const { to, subject, body } = req.body;

    if (!to) {
      res.status(400).json({ error: true, message: 'Recipient email (to) is required' });
      return;
    }

    await sgMail.send({
      to,
      from: process.env.SENDGRID_FROM_EMAIL || 'noreply@seemaai.co.uk',
      subject: subject || 'Test Email from Seema',
      text: body || 'This is a test email sent from the Seema compliance platform.',
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'test_email_sent',
      entityType: 'email',
      metadata: { to, subject },
    });

    res.json({ success: true, message: 'Test email sent' });
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to send test email' });
  }
});

// POST /admin/email/auto-chase — Trigger auto-chase logic
router.post('/admin/email/auto-chase', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const now = new Date();

    // Find chasers that were sent but not acknowledged, and are older than the chase frequency
    const firm = await prisma.firm.findUnique({ where: { id: firmId } });
    let chaseFrequencyDays = 7;

    if (firm?.firmPreferences) {
      try {
        const prefs = JSON.parse(firm.firmPreferences);
        if (prefs.emailSettings?.autoChaseFrequencyDays) {
          chaseFrequencyDays = prefs.emailSettings.autoChaseFrequencyDays;
        }
      } catch { /* use default */ }
    }

    const cutoffDate = new Date(now.getTime() - chaseFrequencyDays * 24 * 60 * 60 * 1000);

    const unresolvedChasers = await prisma.chaserLog.findMany({
      where: {
        firmId,
        status: 'sent',
        sentAt: { lt: cutoffDate },
      },
    });

    let resent = 0;
    for (const chaser of unresolvedChasers) {
      if (!chaser.recipient) continue;

      const newSubject = `Reminder: ${chaser.subject || ''}`;
      try {
        await sgMail.send({
          to: chaser.recipient,
          from: process.env.SENDGRID_FROM_EMAIL || 'noreply@seemaai.co.uk',
          subject: newSubject,
          text: `This is an automated reminder follow-up for: ${chaser.subject}`,
        });

        await prisma.chaserLog.update({
          where: { id: chaser.id },
          data: {
            attempts: (chaser.attempts ?? 0) + 1,
            sentAt: now,
            subject: newSubject,
          },
        });
        resent++;
      } catch { /* skip failed sends */ }
    }

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'auto_chase_triggered',
      entityType: 'chaser_log',
      metadata: { total: unresolvedChasers.length, resent },
    });

    res.json({ total: unresolvedChasers.length, resent });
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to trigger auto-chase' });
  }
});

export default router;
